Changes to GpuEye to enable `k` offset parameter.

shawntan opened the pull request:
The tests will be coming up, thought I'd do a pull request to get some feedback before I carry on.

@ReyhaneAskari #2763

shawntan commented:
- `k` is currently of type `int32`. Should it be `int64`?
- The code currently allows `|k| > max(m, n)`, the result will just be zeroes. Let me know if this is not the desired behaviour
- I've been testing it with the following cases:

```python
import theano
import theano.tensor as T
import numpy as np
n = T.iscalar('n')
m = T.iscalar('m')
k = theano.shared(np.int32(0))

e = T.eye(n, m, k) + 1
f_ = theano.function([n, m], e)


def f(a, b):
    return f_(a, b) - 1

print f(4, 4)
k.set_value(np.int32(1))
print f(4, 4)
k.set_value(np.int32(2))
print f(4, 4)
k.set_value(np.int32(3))
print f(4, 4)
print f(8, 4)
print f(4, 8)
k.set_value(np.int32(-3))
print f(4, 4)
print f(8, 4)
print f(4, 8)
k.set_value(np.int32(-4))
print f(4, 4)
print f(8, 4)
print f(4, 8)
k.set_value(np.int32(4))
print f(4, 4)
print f(8, 4)
print f(4, 8)
```

ReyhaneAskari commented on theano/gpuarray/basic_ops.py:
```diff
@@ -1613,21 +1613,24 @@ def grad(self, inp, grads):
         return [Kernel(
                 code=code, name="eye",
-                params=[gpuarray.GpuArray, gpuarray.SIZE, gpuarray.SIZE],
+                params=[gpuarray.GpuArray, gpuarray.SIZE, gpuarray.SIZE, np.int32],
```
You need to indicate the type of ```k``` as ```'int32'```. Take a look at [this example](https://github.com/Theano/Theano/blob/99e32f41ec92fe2d123c7770074b99674d9d5073/theano/gpuarray/elemwise.py#L2626).

nouiz commented:
I don't think k need to be int64. Don't forget, this is approx the max
number of lines/columns. So an k that don't fit in int32 would be very thin
or tall. But I don't think int64 would hurt.

On Thu, Mar 30, 2017 at 1:55 PM Reyhane Askari <notifications@github.com>
wrote:

> *@ReyhaneAskari* requested changes on this pull request.
>
> Thanks Shawn.
>
>    - I think int64 would be safer. @abergeron
>    <https://github.com/abergeron> what do you think?
>    - As we want it to be consistent with numpy, so yes it is a desired
>    behaviour.
>    - Add another test for k larger than max(m, n) and then add your code
>    in *theano/gpuarray/tests/test_basic_ops.py* and instead of printing
>    add assertions.
>
> ------------------------------
>
> In theano/gpuarray/basic_ops.py
> <https://github.com/Theano/Theano/pull/5789#discussion_r108990940>:
>
> >      }
>  }""" % dict(ctype=pygpu.gpuarray.dtype_to_ctype(self.dtype),
>              name=name, write_a=write_w(self.dtype))
>          return [Kernel(
>                  code=code, name="eye",
> -                params=[gpuarray.GpuArray, gpuarray.SIZE, gpuarray.SIZE],
> +                params=[gpuarray.GpuArray, gpuarray.SIZE, gpuarray.SIZE, np.int32],
>
> You need to indicate the type of k as 'int32'. Take a look at this example
> <https://github.com/Theano/Theano/blob/99e32f41ec92fe2d123c7770074b99674d9d5073/theano/gpuarray/elemwise.py#L2626>
> .
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/Theano/pull/5789#pullrequestreview-30063067>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AALC-ws8qJL6d8UX2UEwY_TC9_Oq76ICks5rq-x0gaJpZM4MuqRU>
> .
>

shawntan commented:
I've added the same test cases I found for the [cpu version](https://github.com/Theano/Theano/blob/master/theano/tensor/tests/test_basic.py) for `T.eye`, but I think in both there should be further test cases where `k > M` or `-k > N` depending on whether `k` is negative.

I've removed the `stack` call, not quite sure why that was there. If there is a reason for it, let me know.

abergeron commented on theano/gpuarray/basic_ops.py:
```diff
@@ -1613,21 +1611,25 @@ def grad(self, inp, grads):
-    ga_size nb = n < m ? n : m;
-    for (ga_size i = LID_0; i < nb; i += LDIM_0) {
-        a[i*m + i] = %(write_a)s(1);
+KERNEL void eye(GLOBAL_MEM %(ctype)s *a, ga_size n, ga_size m, ga_long k) {
```
Use ga_ssize for k to be consistent with the other two args.

shawntan replied:
k needs to be negative, ssize is an unsigned long right?

abergeron replied:
ssize is signed.  size is unsigned.

shawntan replied:
Okay thanks!

abergeron commented on theano/gpuarray/basic_ops.py:
```diff
@@ -1613,21 +1611,25 @@ def grad(self, inp, grads):
+KERNEL void eye(GLOBAL_MEM %(ctype)s *a, ga_size n, ga_size m, ga_long k) {
+    ga_long col_off = max(k, (ga_long) 0);
+    ga_long row_off = -min(k, (ga_long) 0);
+    if (row_off < n && col_off < m) {
```
Instead of doing the check in the kernel, it would probably be a better idea to do this check outside and skip the kernel call completely when the output would be all-zeros.

abergeron replied:
I'm not 100% sure that would work in all cases though.

shawntan replied:
Right.

Do you mean doing the check wouldn't work in all cases, or that the current code as it is wouldn't work in all cases?

abergeron replied:
I mean that moving the check outside might not always work.

ReyhaneAskari commented on theano/gpuarray/basic_ops.py:
```diff
@@ -1674,4 +1679,5 @@ def c_code(self, node, name, inp, out, sub):
     def c_code_cache_version(self):
-        return (7,)
+        # return (7,)
+        return
```
I guess you forgot to remove the return in here.

shawntan replied:
Argh I always forget. Thanks!

ReyhaneAskari commented on theano/gpuarray/basic_ops.py:
```diff
@@ -1658,7 +1658,12 @@ def c_code(self, node, name, inp, out, sub):
+            err = eye_call(1, &gs, &ls, 0, %(z)s->ga.data, dims[0], dims[1], k);
+        }
+
         if (err != GA_NO_ERROR) {
```
I think you will get an error here. You have defined ```err``` but have not initialized it. It will be indeterminate.

shawntan replied:
No error there, because err has already been initialised, probably to the same value as `GA_NO_ERROR`. But you're right it doesn't make sense, I should move it in.

shawntan commented:
Not sure where the flake8 errors are occuring. Running the command locally doesn't reproduce it.

Seems to be happening on other builds too.

abergeron commented:
The flake8 problem is a bug in sphinx on Travis, your code is fine.

nouiz commented:
jenkins test this

nouiz commented on theano/gpuarray/basic_ops.py:
```diff
@@ -1613,21 +1611,23 @@ def grad(self, inp, grads):
 
     def c_code(self, node, name, inp, out, sub):
-        n, m = inp
+        n, m, k = inp
```
Here you should check if inp have 2 or 3 inputs. If only 2, then suppose k is 0 and hardcode zero.

This is to support old pickled files.

lamblin commented:
Jenkins test this please

shawntan commented:
I'm a bit clueless as to how to deal with the errors here.
I've pulled the latest from master and testing them locally on GPU seems to be fine.

ReyhaneAskari commented:
There seems to be a problem either with jenkins or with theano's master that needs to be fixed. Your ticket seems to be complete. Just wait for an update.
Thanks.

nouiz commented:
jenkins test this

nouiz commented:
closed and reopening to trigger the CI with the new base.
